libs/core/langchain_core/indexes/base.py

```diff
@@ -0,0 +1,369 @@
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """Upsert documents to index."""
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async upsertions.
+    async def aupsert(
+        self,
+        documents: AsyncIterable[Document],
+        *,
+        ids: Optional[AsyncIterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """Upsert documents to index. Default implementation, runs sync upsert() in async executor."""
+        # TODO: how to convert AsyncIterable -> Iterable
+        return await run_in_executor(None, self.upsert, documents, ids=ids, **kwargs)
+
+    @abstractmethod
+    def delete_by_ids(self, ids: Iterable[str]) -> DeleteResponse:
+        """Delete documents by id.
+
+        Args:
+            ids: IDs of the documents to delete.
+
+        Returns:
+           A dict ``{"succeeded": [...], "failed": [...]}`` with the IDs of the
+           documents that were successfully deleted and the ones that failed to be
+           deleted.
+        """
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async deletions.
+    async def adelete_by_ids(self, ids: AsyncIterable[str]) -> DeleteResponse:
+        """Upsert documents to index. Default implementation, runs sync delete_by_ids() in async executor."""
+        return await run_in_executor(None, self.delete_by_ids, ids)
+
+    @abstractmethod
+    def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+        """Lazily get documents by id.
+
+        Args:
+            ids: IDs of the documents to get.
+
+        Yields:
+           Document
+        """
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async get.
+    async def alazy_get_by_ids(
+        self, ids: AsyncIterable[str]
+    ) -> AsyncIterable[Document]:
+        """Lazily get documents by id.
+
+        Default implementation, runs sync () in async executor.
+
+        Args:
+            ids: IDs of the documents to get.
+
+        Yields:
+           Document
+        """
+        return await run_in_executor(None, self.lazy_get_by_ids, ids)
+
+    def get_by_ids(self, ids: Iterable[str]) -> List[Document]:
+        """Get documents by id.
+
+        Args:
+            ids: IDs of the documents to get.
+
+        Returns:
+           A list of the requested Documents.
+        """
+        return list(self.lazy_get_by_ids(ids))
+
+    async def aget_by_ids(self, ids: AsyncIterable[str]) -> List[Document]:
+        docs = []
+        async for doc in await self.alazy_get_by_ids(ids):
+            docs.append(doc)
+        return docs
+
+    def delete(
+        self,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        filters: Union[
+            StructuredQuery, Dict[str, Any], List[Dict[str, Any]], None
+        ] = None,
+        **kwargs: Any,
+    ) -> DeleteResponse:
```

Comment:
The `DeleteResponse` doesn't works -- it's impossible to know the IDs if using filters (especially if it fails). 

We should think about the kinds of failure cases that may occur:

1. network connection / bad API key -- most frequent failure
2. less frequent would be an API that raises an exception if the underlying ID doesn't exist -- but this isn't super common